File: precision-sweep-demo/xoshiro.js
// xoshiro128** — must produce byte-identical sequence to precision_sweep/xoshiro.py
(function (root) {
  'use strict';
  const MASK = 0xFFFFFFFF;
  function rotl(x, k) {
    return ((x << k) | (x >>> (32 - k))) >>> 0;
  }
  function mul32(a, b) {
    const aHi = (a >>> 16);
    const aLo = a & 0xFFFF;
    const bHi = (b >>> 16);
    const bLo = b & 0xFFFF;
    return (((aHi * bLo + aLo * bHi) << 16) + aLo * bLo) >>> 0;
  }
  function Xoshiro128SS(seed) {
    let s = (seed >>> 0);
    const state = [];
    for (let i = 0; i < 4; i++) {
      s = (s + 0x9E3779B9) >>> 0;
      let z = s;
      z = mul32(z ^ (z >>> 16), 0x85EBCA6B);
      z = mul32(z ^ (z >>> 13), 0xC2B2AE35);
      z = (z ^ (z >>> 16)) >>> 0;
      state.push(z);
    }
    if (state.every(x => x === 0)) { state[0] = 1; }
    this.s = state;
  }
  Xoshiro128SS.prototype.nextUint32 = function () {
    const result = (mul32(rotl(mul32(this.s[1], 5), 7), 9)) >>> 0;
    const t = (this.s[1] << 9) >>> 0;
    this.s[2] ^= this.s[0];
    this.s[3] ^= this.s[1];
    this.s[1] ^= this.s[2];
    this.s[0] ^= this.s[3];
    this.s[2] ^= t;
    this.s[3] = rotl(this.s[3], 11);
    return result;
  };
  Xoshiro128SS.prototype.random = function () {
    return this.nextUint32() / 4294967296.0;
  };
  Xoshiro128SS.prototype.uniform = function (lo, hi) {
    return lo + (hi - lo) * this.random();
  };
  Xoshiro128SS.prototype.normal = function (mu, sigma) {
    let u1 = this.random();
    const u2 = this.random();
    if (u1 === 0) u1 = 1e-12;
    const z = Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);
    return mu + sigma * z;
  };
  root.Xoshiro128SS = Xoshiro128SS;
})(typeof window !== 'undefined' ? window : globalThis);

File: precision_sweep/xoshiro.py
"""xoshiro128** PRNG — deterministic across Python and JS.

Reference: https://prng.di.unimi.it/xoshiro128starstar.c
32-bit state, period 2^128 - 1, passes PractRand to 32 TB.
"""
from __future__ import annotations


MASK = 0xFFFFFFFF


def _rotl(x: int, k: int) -> int:
    return ((x << k) | (x >> (32 - k))) & MASK


class Xoshiro128SS:
    """32-bit xoshiro128** generator. Seeds from a single integer via
    SplitMix32-style expansion to four 32-bit words."""

    def __init__(self, seed: int):
        if not isinstance(seed, int) or seed < 0:
            raise ValueError("seed must be non-negative int")
        # Expand a single int to 4x 32-bit state via SplitMix32
        s = seed & MASK
        self.s = []
        for _ in range(4):
            s = (s + 0x9E3779B9) & MASK
            z = s
            z = ((z ^ (z >> 16)) * 0x85EBCA6B) & MASK
            z = ((z ^ (z >> 13)) * 0xC2B2AE35) & MASK
            z = (z ^ (z >> 16)) & MASK
            self.s.append(z)
        # Guard against all-zero state (xoshiro forbids it)
        if all(x == 0 for x in self.s):
            self.s = [1, 0, 0, 0]

    def next_uint32(self) -> int:
        result = (_rotl((self.s[1] * 5) & MASK, 7) * 9) & MASK
        t = (self.s[1] << 9) & MASK
        self.s[2] ^= self.s[0]
        self.s[3] ^= self.s[1]
        self.s[1] ^= self.s[2]
        self.s[0] ^= self.s[3]
        self.s[2] ^= t
        self.s[3] = _rotl(self.s[3], 11)
        return result

    def random(self) -> float:
        """Return a float in [0, 1) with 24 bits of precision."""
        return self.next_uint32() / 4294967296.0

    def uniform(self, lo: float, hi: float) -> float:
        return lo + (hi - lo) * self.random()

    def normal(self, mu: float, sigma: float) -> float:
        """Box-Muller: two uniforms → one normal. Deterministic."""
        import math
        u1 = self.random()
        u2 = self.random()
        # Guard u1 > 0 to avoid log(0); PRNG output is [0,1) so u1 can be 0
        if u1 == 0.0:
            u1 = 1e-12
        z = math.sqrt(-2.0 * math.log(u1)) * math.cos(2.0 * math.pi * u2)
        return mu + sigma * z
